pkg/query/quick_e2e_test.go

```diff
@@ -0,0 +1,115 @@
+package query_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/authzed/spicedb/internal/datastore/dsfortesting"
+	"github.com/authzed/spicedb/internal/datastore/memdb"
+	"github.com/authzed/spicedb/internal/testfixtures"
+	corev1 "github.com/authzed/spicedb/pkg/proto/core/v1"
+	"github.com/authzed/spicedb/pkg/query"
+	"github.com/authzed/spicedb/pkg/schema/v2"
+)
+
+func TestCheck(t *testing.T) {
+	t.Parallel()
+
+	require := require.New(t)
+	rawDS, err := dsfortesting.NewMemDBDatastoreForTesting(0, 0, memdb.DisableGC)
+	require.NoError(err)
+
+	ds, revision := testfixtures.StandardDatastoreWithData(rawDS, require)
+
+	// This stands in for the step of fetching and caching the schema locally.
+	objectDefs := []*corev1.NamespaceDefinition{testfixtures.UserNS.CloneVT(), testfixtures.FolderNS.CloneVT(), testfixtures.DocumentNS.CloneVT()}
+	dsSchema, err := schema.BuildSchemaFromDefinitions(objectDefs, nil)
+	require.NoError(err)
+
+	// This stands in for the step that pre-builds query trees from the schema,
+	// by iterating through the relationships in the schema and then walking them.
+	//
+	// In this case, it's a little contrived.
+	vande := query.NewRelationIterator(dsSchema.Definitions["document"].Relations["viewer_and_editor"].BaseRelations[0])
+	edit := query.NewRelationIterator(dsSchema.Definitions["document"].Relations["editor"].BaseRelations[0])
+	it := query.NewIntersection()
+	it.AddSubIterator(vande)
+	it.AddSubIterator(edit)
```

Comment:
I wonder if we should have `AddSubIterator` return an *updated* struct and keep all of this as edit-by-copy: its likely worse for mem usage, but far easier to reason about?

Reply:
That's a thought, and I've done that in Go code before in other circumstances (mostly when it's something like `type Foo uint64` and I'm tweaking the bits) -- here, building is, aside from schema read time, kinda rare, so I'm not concerned.

One compromise I kinda like and will go ahead and do preemptively is to simply make AddSubIterator private -- only this package can mutate like that, nothing else.